plugin/documents.py

```diff
@@ -471,17 +474,33 @@ def on_hover(self, point: int, hover_zone: int) -> None:
         if hover_zone == sublime.HOVER_TEXT and window and window.settings().get(HOVER_ENABLED_KEY, True):
             self.view.run_command("lsp_hover", {"point": point})
         elif hover_zone == sublime.HOVER_GUTTER:
-            # Lightbulb must be visible and at the same line
-            if self._lightbulb_line != self.view.rowcol(point)[0]:
-                return
-            content = code_actions_content(self._actions_by_config)
-            if content:
-                show_lsp_popup(
-                    self.view,
-                    content,
-                    flags=sublime.HIDE_ON_MOUSE_MOVE_AWAY,
-                    location=point,
-                    on_navigate=lambda href: self._on_navigate(href, point))
+            sublime.set_timeout_async(partial(self._on_hover_gutter_async, point))
+
+    def _on_hover_gutter_async(self, point: int) -> None:
+        content = ''
+        if self._lightbulb_line == self.view.rowcol(point)[0]:
+            content += code_actions_content(self._actions_by_config)
+        if userprefs().diagnostics_gutter_marker and userprefs().show_diagnostics_severity_level:
```

Comment:
Is this the expected behavior?

1 
`"diagnostics_gutter_marker": "circle",`
<img width="526" alt="Screenshot 2023-10-31 at 22 59 16" src="https://github.com/sublimelsp/LSP/assets/115091796/428d04b2-4aad-4284-b62d-ce7519badbbc">

If there are 2 diagnostics, 
and `diagnostics_gutter_marker` is set to `"circle"`
**display** the diagnostics on gutter hover.

2
`"diagnostics_gutter_marker": "",`
<img width="518" alt="Screenshot 2023-10-31 at 22 58 53" src="https://github.com/sublimelsp/LSP/assets/115091796/ca31dd10-54f9-4ba1-9ad2-82327c1bd1a7">

If there are 2 diagnostics, 
and `diagnostics_gutter_marker` is set to `""`
**don't display** the diagnostics on gutter hover.

Reply:
In case 2 it doesn't show the diagnostics in addition to the code actions, because in general they can be unrelated to each other. But I'm not sure, maybe would be better to also see the diagnostics, what do you think? In that case it should probably only filter for the diagnostics which are overlapping the current caret position or selection.

Reply:
If I did it correctly, then the popup in your example 2 should show only the first diagnostic with small "d" now (where the caret is).

Reply:
> If I did it correctly, then the popup in your example 2 should show only the first diagnostic with small "d" now (where the caret is).

That is correct.
I will see the diagnotics where the cursor is if I hover the gutter.

<img width="581" alt="Screenshot 2023-11-04 at 22 37 35" src="https://github.com/sublimelsp/LSP/assets/115091796/7446ac1f-fb4a-4b8c-83e6-d52da37aa5c3">

But I am not sure I think that behavior is the best choice, 
because I can already hover with the cursor over the diagnostics and see the diagnostic.
So it is kind of redudant.

>  it should probably only filter for the diagnostics which are overlapping the current caret position or selection.

For me it would be more intuitive 
if hovering the gutter should show all diagnostics that **just exist on the line**. (not based on cursor or selection)

So my expected behavior would be the following:
if `"diagnostics_gutter_marker": ""`
If there are 2 diagnostics,
display the 2 diagnostics on gutter hover.

Reply:
> But I am not sure I think that behavior is the best choice,
> because I can already hover with the cursor over the diagnostics and see the diagnostic.
> So it is kind of redudant.

In this case with the diagnostic icons disabled it will show only the diagnostic(s) which is/are sent as part of the code action request ([CodeActionContext](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeActionContext)). So there is a good chance that the code actions are related to the diagnostics at the cursor position, especially because we only show Quick Fix actions here. And the code actions request is sent for a particular `range`, so the actions should also be only applicable for the current cursor position.

You could consider the diagnostic from this screenshot to be part of the lightbulb hover, and the other diagnostics on the line or on different lines are not shown because the diagnostics gutter icons are disabled. I think it makes sense.

And I would not say that it is redundant, it's rather just a different way to see the diagnostic message. For example I have the code actions in the regular hover popup disabled, so I will never see them if I hover over the error underline.